server/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -830,7 +830,11 @@ private void sync(final Consumer<IndexShard> sync, final String source) {
                     case STARTED:
                         try {
                             shard.runUnderPrimaryPermit(
```

Comment:
I think this could potentially still fail if the shardRouting is updated between checking in line 821 and asserting inside `runUnderPrimaryPermit`. At least I cannot find the protection against that. If others agree, this could be something for a follow-up.

Reply:
Once the shardRouting is an active primary, it will never be anything else. We never move a shard from primary to replica, and never move a shard from active to non-active. In each of these cases, the existing IndexShard instance will be remove and a new one created.